doc/users-guide/datapar/promotion.rst

```diff
@@ -0,0 +1,324 @@
+Promoted calls do not support a mechanism to override default forall
+intents.  Thus, in order to override the default and enable the race,
+a user would need to write the loop out explicitly and add a ``with``
+clause as follows:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.chpl
+  :language: chapel
+  :lines: 114-115
+
+
+Promoting Using Ranges and Domains
+----------------------------------
+
+As mentioned at the outset, not only can arrays be used to promote
+functions, but ranges and domains can as well.  For example, the
+following two calls promote ``maybeCopy()`` using a range and a
+domain, respectively:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.chpl
+  :language: chapel
+  :lines: 185,189
+
+The contents of *A* after these calls is as follows:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.good
+  :language: text
+  :lines: 45,46
+
+At the time of this writing, Chapel does not support an official way
+for users to create their own collection types that support promotion.
+We expect to support this in the future by having the collection type
+support certain well-defined iterator methods.
+
+
+Promoted Library Routines and Operators
+---------------------------------------
+
+All of the examples above illustrate promotion on procedures written
+by a user.  Yet any procedure is a candidate for promotion in Chapel,
+including library routines and operators.  As an example, consider the
+following lines of code:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.chpl
+  :language: chapel
+  :lines: 119-120,127
+
+The second statement promotes the standard ``sqrt()`` library routine,
+resulting in an array of results, stored in *B*:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.good
+  :language: text
+  :lines: 32
+
+The third statement doubles the array *A* and stores the result in *B*
+as expected:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.good
+  :language: text
+  :lines: 34
+
+Note that Chapel does not support an explicit ``+`` operator for
+arrays of reals.  Rather, Chapel implements this operation by
+promoting the standard ``+`` operator for scalar values of type
+``real``.  For this reason, applying ``*`` to Chapel arrays results in
+an elementwise multiplication of the arrays' elements by default,
+rather than a matrix multiplication operation.
+
+Even the assignment operations in these statements can be considered
+to be promotions of scalar assignment for ``real`` values.  Thus, the
+statements above can be considered to be equivalent to:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.chpl
+  :language: chapel
+  :lines: 123-124,130-131
+
+
+Promotion vs. Whole-Array Operations
+------------------------------------
+
+A key distinction should be made between Chapel's use of promotion
+versus whole-array operations as supported in traditional array
+languages.  To see the difference, consider the following statement:
```

Comment:
When I initially read this paragraph, I thought the following would expect me to know something about "traditional array languages". It does not - in fact it explains that context very well. However there might be something that can be done here to assure the reader that this is still introductory material...